ntoskrnl/ke/i386/mproc.c

```diff
@@ -0,0 +1,148 @@
+        if (!APInfo)
+        {
+            break;
+        }
+        KernelStack = MmCreateKernelStack(FALSE, 0);
+        if (!KernelStack)
+        {
+            break;
+        }
+        DPCStack = MmCreateKernelStack(FALSE, 0);
+        if (!DPCStack)
+        {
+            break;
+        }
+
+        /* Initalize a new PCR for the specific AP */
+        KiInitializePcr(ProcessorCount,
+                        &APInfo->Pcr,
+                        &APInfo->Idt[0], 
+                        &APInfo->Gdt[0], 
+                        &APInfo->Tss,
+                        (PKTHREAD)&APInfo->Thread,
+                        DPCStack);
+
+        // Prepare descriptor tables
+        KDESCRIPTOR bspGdt, bspIdt;
+        __sgdt(&bspGdt.Limit);
+        __sidt(&bspIdt.Limit);
+        RtlCopyMemory(&APInfo->Gdt, (PVOID)bspGdt.Base, bspGdt.Limit + 1);
+        RtlCopyMemory(&APInfo->Idt, (PVOID)bspIdt.Base, bspIdt.Limit + 1);
+
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_R0_PCR), (ULONG_PTR)&APInfo->Pcr);
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_DF_TSS), (ULONG_PTR)&APInfo->TssDoubleFault);
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_NMI_TSS), (ULONG_PTR)&APInfo->TssNMI);
+
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_TSS), (ULONG_PTR)&APInfo->Tss);
+        // Clear TSS Busy flag (aka set the type to "TSS (Available)")
+        KiGetGdtEntry(&APInfo->Gdt, KGDT_TSS)->HighWord.Bits.Type = 0b1001;
+
+        APInfo->TssDoubleFault.Esp0 = (ULONG_PTR)&APInfo->NMIStackData;
+        APInfo->TssDoubleFault.Esp = (ULONG_PTR)&APInfo->NMIStackData;
+
+        APInfo->TssNMI.Esp0 = (ULONG_PTR)&APInfo->NMIStackData;
+        APInfo->TssNMI.Esp = (ULONG_PTR)&APInfo->NMIStackData;
+
+        // Fill the processor state
+        PKPROCESSOR_STATE ProcessorState = &APInfo->Pcr.Prcb->ProcessorState;
+        RtlZeroMemory(ProcessorState, sizeof(*ProcessorState));
+
+        ProcessorState->SpecialRegisters.Cr0 = __readcr0();
+        ProcessorState->SpecialRegisters.Cr3 = __readcr3();
+        ProcessorState->SpecialRegisters.Cr4 = __readcr4();
+
+        ProcessorState->ContextFrame.SegCs = KGDT_R0_CODE;
+        ProcessorState->ContextFrame.SegDs = KGDT_R3_DATA;
+        ProcessorState->ContextFrame.SegEs = KGDT_R3_DATA;
+        ProcessorState->ContextFrame.SegSs = KGDT_R0_DATA;
+        ProcessorState->ContextFrame.SegFs = KGDT_R0_PCR;
+
+        ProcessorState->SpecialRegisters.Gdtr.Base = (ULONG_PTR)APInfo->Gdt;
+        ProcessorState->SpecialRegisters.Gdtr.Limit = sizeof(APInfo->Gdt) - 1;
+        ProcessorState->SpecialRegisters.Idtr.Base = (ULONG_PTR)APInfo->Idt;
+        ProcessorState->SpecialRegisters.Idtr.Limit = sizeof(APInfo->Idt) - 1;
+
+        ProcessorState->SpecialRegisters.Tr = KGDT_TSS;
+
+        ProcessorState->ContextFrame.Esp = (ULONG_PTR)KernelStack;
+        ProcessorState->ContextFrame.Eip = (ULONG_PTR)KiSystemStartup;
```

Comment:
On Windows this is done more logically in this KeStartAllProcessors function itself, not in the hal at all.
(Your current design is not that great because in one place you specify which function is going to be called, but without its arguments, and then in a completely different place you setup the parameters of this function, but without specifying the function itself... The HAL shouldn't have to "know" what the kernel itself wants the CPU to run.)

Reply:
https://github.com/reactos/reactos/pull/5879#discussion_r1381916734

kinda sounds like you're both saying directly opposite things here. 
I don't mind either way but i can't address both of these comments :)

Reply:
I think that both are correct. The template stub code is set by the HAL, and is independent of which kernel entry-point is going to be started.
Separately, the kernel itself knows what kernel entrypoint it wants the CPU to run (which function, with which parameters).

Reply:
Disregard my earlier comment about moving this to the hal. I agree with Hermes.

Reply:
Alright then it's decided it's moved back into mproc.c :)